File: elixir/lib/symphony/orchestrator.ex
defmodule Symphony.Orchestrator do
  @moduledoc """
  Poll-dispatch-monitor GenServer.
  Periodically fetches issues from the tracker, dispatches agent runs,
  handles retries, stall detection, and reconciliation.
  """
  use GenServer
  require Logger

  alias Symphony.{Config, AgentRunner}

  defstruct [
    :tracker,
    :bridge,
    :workflow,
    :hooks,
    :task_supervisor,
    running: %{},
    completed: MapSet.new(),
    claimed: MapSet.new(),
    retry_queue: [],
    token_totals: %{cost_usd: 0.0, turns: 0},
    monitors: %{}
  ]

  # --- Public API ---

  def start_link(opts \\ []) do
    name = Keyword.get(opts, :name, __MODULE__)
    GenServer.start_link(__MODULE__, opts, name: name)
  end

  def state(name \\ __MODULE__) do
    GenServer.call(name, :get_state)
  end

  def refresh(name \\ __MODULE__) do
    send(name, :tick)
    :ok
  end

  # --- Callbacks ---

  @impl true
  def init(opts) do
    tracker = Keyword.get(opts, :tracker, Symphony.GitHub.Adapter)
    bridge = Keyword.get(opts, :bridge, Symphony.Claude.NodeBridge)
    workflow = Keyword.get(opts, :workflow)
    hooks = Keyword.get(opts, :hooks, %{})
    task_supervisor = Keyword.get(opts, :task_supervisor, Symphony.TaskSupervisor)

    state = %__MODULE__{
      tracker: tracker,
      bridge: bridge,
      workflow: workflow,
      hooks: hooks,
      task_supervisor: task_supervisor
    }

    schedule_tick(Config.poll_interval_ms())
    {:ok, state}
  end

  @impl true
  def handle_call(:get_state, _from, state) do
    snapshot = %{
      running: Map.keys(state.running),
      completed: MapSet.to_list(state.completed),
      claimed: MapSet.to_list(state.claimed),
      retry_queue: Enum.map(state.retry_queue, fn {id, attempt, _} -> {id, attempt} end),
      token_totals: state.token_totals
    }

    {:reply, snapshot, state}
  end

  @impl true
  def handle_info(:tick, state) do
    state = poll_and_dispatch(state)
    state = check_stalls(state)
    state = process_retries(state)
    schedule_tick(Config.poll_interval_ms())
    {:noreply, state}
  end

  # Task completion
  def handle_info({ref, {:ok, result}}, state) when is_reference(ref) do
    Process.demonitor(ref, [:flush])
    state = handle_task_success(ref, result, state)
    {:noreply, state}
  end

  def handle_info({ref, {:error, reason}}, state) when is_reference(ref) do
    Process.demonitor(ref, [:flush])
    state = handle_task_failure(ref, reason, state)
    {:noreply, state}
  end

  # Task crash
  def handle_info({:DOWN, ref, :process, _pid, reason}, state) do
    state = handle_task_failure(ref, {:crash, reason}, state)
    {:noreply, state}
  end

  def handle_info(_msg, state), do: {:noreply, state}

  @impl true
  def terminate(_reason, state) do
    Logger.info("[Orchestrator] Shutting down, stopping #{map_size(state.running)} running agents")

    for {_id, %{task: task}} <- state.running do
      Task.Supervisor.terminate_child(state.task_supervisor, task.pid)
    end

    :ok
  end

  # --- Private ---

  defp poll_and_dispatch(state) do
    case state.tracker.list_issues() do
      {:ok, issues} ->
        candidates = filter_candidates(issues, state)
        dispatch_candidates(candidates, state)

      {:error, reason} ->
        Logger.error("[Orchestrator] Poll failed: #{inspect(reason)}")
        state
    end
  end

  defp filter_candidates(issues, state) do
    running_ids = MapSet.new(Map.keys(state.running))
    all_issue_numbers = MapSet.new(Enum.map(issues, &Symphony.GitHub.Issue.identifier/1))

    issues
    |> Enum.filter(fn issue ->
      id = Symphony.GitHub.Issue.identifier(issue)

      not MapSet.member?(running_ids, id) and
        not MapSet.member?(state.completed, id) and
        not MapSet.member?(state.claimed, id) and
        issue.state == "open" and
        blockers_resolved?(issue, all_issue_numbers, state)
    end)
    |> Enum.sort_by(&priority_sort_key/1)
  end

  defp blockers_resolved?(issue, _all_ids, state) do
    Enum.all?(issue.blockers, fn blocker_num ->
      blocker_id = to_string(blocker_num)
      MapSet.member?(state.completed, blocker_id)
    end)
  end

  defp priority_sort_key(issue) do
    case issue.priority do
      :p1 -> 1
      :p2 -> 2
      :p3 -> 3
      :p4 -> 4
    end
  end

  defp dispatch_candidates(candidates, state) do
    slots = Config.max_concurrent() - map_size(state.running)

    candidates
    |> Enum.take(max(slots, 0))
    |> Enum.reduce(state, &dispatch_issue/2)
  end

  defp dispatch_issue(issue, state) do
    identifier = Symphony.GitHub.Issue.identifier(issue)
    Logger.info("[Orchestrator] Dispatching issue ##{identifier}: #{issue.title}")

    task =
      Task.Supervisor.async_nolink(state.task_supervisor, fn ->
        AgentRunner.run(issue,
          bridge: state.bridge,
          tracker: state.tracker,
          hooks: state.hooks,
          workflow: state.workflow
        )
      end)

    running =
      Map.put(state.running, identifier, %{
        task: task,
        issue: issue,
        started_at: System.monotonic_time(:millisecond),
        last_event_at: System.monotonic_time(:millisecond)
      })

    claimed = MapSet.put(state.claimed, identifier)

    broadcast(:issue_dispatched, %{issue_number: identifier, title: issue.title})

    %{state | running: running, claimed: claimed}
  end

  defp handle_task_success(ref, result, state) do
    case find_by_ref(ref, state) do
      {identifier, _info} ->
        running = Map.delete(state.running, identifier)
        completed = MapSet.put(state.completed, identifier)

        cost = result[:cost_usd] || 0
        turns = result[:num_turns] || 0

        totals = %{
          cost_usd: state.token_totals.cost_usd + cost,
          turns: state.token_totals.turns + turns
        }

        broadcast(:issue_completed, %{issue_number: identifier, cost_usd: cost, turns: turns})

        %{state | running: running, completed: completed, token_totals: totals}

      nil ->
        state
    end
  end

  defp handle_task_failure(ref, reason, state) do
    case find_by_ref(ref, state) do
      {identifier, _info} ->
        Logger.error("[Orchestrator] Issue ##{identifier} failed: #{inspect(reason)}")
        running = Map.delete(state.running, identifier)

        # Schedule retry with exponential backoff
        attempt = retry_attempt(identifier, state) + 1
        max_backoff = Config.max_retry_backoff_ms()
        backoff = min(round(:math.pow(2, attempt) * 1_000), max_backoff)
        retry_at = System.monotonic_time(:millisecond) + backoff

        retry_queue = state.retry_queue ++ [{identifier, attempt, retry_at}]

        broadcast(:issue_failed, %{
          issue_number: identifier,
          error: inspect(reason),
          retry_attempt: attempt,
          retry_in_ms: backoff
        })

        %{state | running: running, retry_queue: retry_queue}

      nil ->
        state
    end
  end

  defp check_stalls(state) do
    now = System.monotonic_time(:millisecond)
    stall_timeout = Config.stall_timeout_ms()

    stalled =
      Enum.filter(state.running, fn {_id, info} ->
        now - info.last_event_at > stall_timeout
      end)

    Enum.reduce(stalled, state, fn {identifier, info}, acc ->
      Logger.warning("[Orchestrator] Issue ##{identifier} stalled, killing")
      Task.Supervisor.terminate_child(acc.task_supervisor, info.task.pid)
      running = Map.delete(acc.running, identifier)
      %{acc | running: running}
    end)
  end

  defp process_retries(state) do
    now = System.monotonic_time(:millisecond)

    {ready, pending} =
      Enum.split_with(state.retry_queue, fn {_, _, retry_at} -> retry_at <= now end)

    state = %{state | retry_queue: pending}

    Enum.reduce(ready, state, fn {identifier, _attempt, _}, acc ->
      # Re-fetch the issue to check if still open
      case acc.tracker.get_issue(identifier) do
        {:ok, issue} when issue.state == "open" ->
          # Remove from claimed so it can be redispatched
          claimed = MapSet.delete(acc.claimed, identifier)
          %{acc | claimed: claimed}

        _ ->
          Logger.info("[Orchestrator] Issue ##{identifier} no longer open, skipping retry")
          acc
      end
    end)
  end

  defp find_by_ref(ref, state) do
    Enum.find(state.running, fn {_id, info} -> info.task.ref == ref end)
    |> case do
      {id, info} -> {id, info}
      nil -> nil
    end
  end

  defp retry_attempt(identifier, state) do
    case Enum.find(state.retry_queue, fn {id, _, _} -> id == identifier end) do
      {_, attempt, _} -> attempt
      nil -> 0
    end
  end

  defp schedule_tick(interval) do
    Process.send_after(self(), :tick, interval)
  end

  defp broadcast(event, payload) do
    Phoenix.PubSub.broadcast(
      Symphony.PubSub,
      "symphony:events",
      {event, payload}
    )
  end
end

File: src/orchestrator.ts
import { Effect, Ref, PubSub, Fiber, Schedule, Duration, Layer } from "effect"
import { Config, ConfigLive, type SymphonyConfig } from "./config.js"
import { EventBus, EventBusLive, type SymphonyEvent } from "./event-bus.js"
import { TrackerService, TrackerLive, type Tracker } from "./github/tracker.js"
import { type Issue, identifier } from "./github/issue.js"
import { type Workflow } from "./workflow.js"
import { runAgent } from "./agent-runner.js"
import { removeWorkspace, sanitizeKey } from "./workspace.js"
import { join } from "node:path"

// --- Domain types ---

interface RunningTask {
  readonly fiber: Fiber.RuntimeFiber<unknown, unknown>
  readonly issue: Issue
  readonly startedAt: number
  readonly lastEventAt: number
  readonly attempt: number | null
}

interface RetryEntry {
  readonly id: string
  readonly identifier: string
  readonly attempt: number
  readonly retryAt: number
  readonly error: string | null
  readonly isContinuation: boolean
}

export interface OrchestratorState {
  readonly running: Map<string, RunningTask>
  readonly completed: Set<string>        // bookkeeping only — does NOT gate dispatch
  readonly claimed: Set<string>
  readonly retryQueue: RetryEntry[]
  readonly tokenTotals: { costUsd: number; turns: number }
}

const initialState: OrchestratorState = {
  running: new Map(),
  completed: new Set(),
  claimed: new Set(),
  retryQueue: [],
  tokenTotals: { costUsd: 0, turns: 0 },
}

export interface OrchestratorHandle {
  readonly state: Ref.Ref<OrchestratorState>
  readonly fiber: Fiber.RuntimeFiber<void, never>
  readonly refresh: () => Effect.Effect<void>
}

/** Start the orchestrator loop. Returns a handle for state access and control. */
export function startOrchestrator(
  workflow: Workflow | null,
  hooks: Record<string, string> = {},
): Effect.Effect<OrchestratorHandle, never, Config | EventBus | TrackerService> {
  return Effect.gen(function* () {
    const config = yield* Config
    const pubsub = yield* EventBus
    const tracker = yield* TrackerService
    const stateRef = yield* Ref.make(initialState)

    const agentLayer = Layer.mergeAll(
      ConfigLive(config),
      Layer.succeed(EventBus, pubsub),
      TrackerLive.pipe(Layer.provide(ConfigLive(config))),
    )

    const tick = Effect.gen(function* () {
      yield* reconcileRunningIssues(stateRef, config, pubsub, tracker, hooks)
      yield* pollAndDispatch(stateRef, config, pubsub, tracker, workflow, hooks, agentLayer)
      yield* processRetries(stateRef, config, pubsub, tracker, workflow, hooks, agentLayer)
    }).pipe(Effect.catchAllCause((cause) => Effect.logError(`Orchestrator tick failed: ${cause}`)))

    const loop = tick.pipe(
      Effect.repeat(Schedule.spaced(Duration.millis(config.pollIntervalMs))),
      Effect.asVoid,
    )

    const fiber = yield* Effect.fork(loop)

    return { state: stateRef, fiber, refresh: () => tick } satisfies OrchestratorHandle
  })
}

/** Get a snapshot of orchestrator state for API/dashboard. */
export function getStateSnapshot(stateRef: Ref.Ref<OrchestratorState>) {
  return Ref.get(stateRef).pipe(
    Effect.map((s) => ({
      running: Array.from(s.running.entries()).map(([id, task]) => ({
        id,
        identifier: identifier(task.issue),
        title: task.issue.title,
        state: task.issue.state,
        startedAt: new Date(task.startedAt).toISOString(),
        lastEventAt: new Date(task.lastEventAt).toISOString(),
        attempt: task.attempt,
      })),
      completed: Array.from(s.completed),
      retryQueue: s.retryQueue.map(({ id, identifier, attempt, retryAt, error, isContinuation }) => ({
        id,
        identifier,
        attempt,
        dueAt: new Date(retryAt).toISOString(),
        error,
        isContinuation,
      })),
      tokenTotals: s.tokenTotals,
      repo: "",
    })),
  )
}

// --- Internal ---

function isActiveState(state: string, config: SymphonyConfig): boolean {
  const normalized = state.toLowerCase()
  return config.trackerActiveStates.some((s) => s.toLowerCase() === normalized)
}

function isTerminalState(state: string, config: SymphonyConfig): boolean {
  const normalized = state.toLowerCase()
  return config.trackerTerminalStates.some((s) => s.toLowerCase() === normalized)
}

function filterCandidates(issues: Issue[], state: OrchestratorState, config: SymphonyConfig): Issue[] {
  const runningIds = new Set(state.running.keys())

  return issues
    .filter((issue) => {
      const id = identifier(issue)
      if (runningIds.has(id)) return false
      if (state.claimed.has(id)) return false

      // Must be in active state and not terminal
      if (!isActiveState(issue.state, config)) return false
      if (isTerminalState(issue.state, config)) return false

      // Blocker rule: if issue state is "open" (todo-equivalent), block if any blocker is non-terminal
      if (issue.state.toLowerCase() === "open") {
        const hasNonTerminalBlocker = issue.blockers.some(
          (b) => !state.completed.has(String(b)),
        )
        if (hasNonTerminalBlocker) return false
      }

      return true
    })
    .sort((a, b) => {
      // 1. Priority ascending (lower number = higher priority, null sorts last)
      const pa = priorityOrder(a.priority)
      const pb = priorityOrder(b.priority)
      if (pa !== pb) return pa - pb

      // 2. created_at oldest first
      const ca = new Date(a.createdAt).getTime()
      const cb = new Date(b.createdAt).getTime()
      if (ca !== cb) return ca - cb

      // 3. identifier lexicographic tie-breaker
      return identifier(a).localeCompare(identifier(b))
    })
}

function priorityOrder(p: string): number {
  switch (p) {
    case "p1": return 1
    case "p2": return 2
    case "p3": return 3
    case "p4": return 4
    default: return 99 // null/unknown sorts last per spec
  }
}

// --- Reconciliation (Section 8.5) ---

function reconcileRunningIssues(
  stateRef: Ref.Ref<OrchestratorState>,
  config: SymphonyConfig,
  pubsub: PubSub.PubSub<SymphonyEvent>,
  tracker: Tracker,
  hooks: Record<string, string>,
): Effect.Effect<void> {
  return Effect.gen(function* () {
    const now = Date.now()
    const state = yield* Ref.get(stateRef)

    if (state.running.size === 0) return

    // Part A: Stall detection
    if (config.stallTimeoutMs > 0) {
      for (const [id, task] of state.running) {
        const lastActivity = task.lastEventAt || task.startedAt
        if (now - lastActivity > config.stallTimeoutMs) {
          yield* Effect.logWarning(`Issue #${id} stalled, interrupting`)
          yield* Fiber.interrupt(task.fiber)
          yield* PubSub.publish(pubsub, {
            _tag: "IssueStalled",
            issueNumber: id,
          } satisfies SymphonyEvent)
          yield* Ref.update(stateRef, (s) => {
            const running = new Map(s.running)
            running.delete(id)
            // Keep claimed so retry can re-dispatch; retry handler will release if ineligible
            const retryQueue = [...s.retryQueue, {
              id,
              identifier: identifier(task.issue),
              attempt: (task.attempt ?? 0) + 1,
              retryAt: now + computeBackoff((task.attempt ?? 0) + 1, config.maxRetryBackoffMs),
              error: "stalled",
              isContinuation: false,
            }]
            return { ...s, running, retryQueue }
          })
        }
      }
    }

    // Part B: Tracker state refresh
    const runningState = yield* Ref.get(stateRef)
    const runningIds = Array.from(runningState.running.keys())
    if (runningIds.length === 0) return

    const refreshResult = yield* Effect.all(
      runningIds.map((id) =>
        tracker.getIssue(id).pipe(
          Effect.map((issue) => ({ id, issue, ok: true as const })),
          Effect.catchAll(() => Effect.succeed({ id, issue: null, ok: false as const })),
        ),
      ),
    )

    for (const { id, issue, ok } of refreshResult) {
      if (!ok) continue // Refresh failed — keep worker running, try next tick

      if (!issue) continue

      if (isTerminalState(issue.state, config)) {
        // Terminal: stop worker and clean workspace
        const task = runningState.running.get(id)
        if (task) {
          yield* Fiber.interrupt(task.fiber)
          const wsDir = join(config.workspaceRoot, sanitizeKey(id))
          yield* removeWorkspace(wsDir, hooks, config.hookTimeoutMs).pipe(
            Effect.catchAll(() => Effect.void),
          )
          yield* Ref.update(stateRef, (s) => {
            const running = new Map(s.running)
            running.delete(id)
            const claimed = new Set(s.claimed)
            claimed.delete(id)
            return { ...s, running, claimed }
          })
        }
      } else if (isActiveState(issue.state, config)) {
        // Still active: update the in-memory issue snapshot
        yield* Ref.update(stateRef, (s) => {
          const running = new Map(s.running)
          const existing = running.get(id)
          if (existing) {
            running.set(id, { ...existing, issue })
          }
          return { ...s, running }
        })
      } else {
        // Neither active nor terminal: stop without workspace cleanup
        const task = runningState.running.get(id)
        if (task) {
          yield* Fiber.interrupt(task.fiber)
          yield* Ref.update(stateRef, (s) => {
            const running = new Map(s.running)
            running.delete(id)
            const claimed = new Set(s.claimed)
            claimed.delete(id)
            return { ...s, running, claimed }
          })
        }
      }
    }
  })
}

// --- Poll and Dispatch (Section 8.1-8.2) ---

function pollAndDispatch(
  stateRef: Ref.Ref<OrchestratorState>,
  config: SymphonyConfig,
  pubsub: PubSub.PubSub<SymphonyEvent>,
  tracker: Tracker,
  workflow: Workflow | null,
  hooks: Record<string, string>,
  agentLayer: Layer.Layer<Config | EventBus | TrackerService>,
): Effect.Effect<void> {
  return Effect.gen(function* () {
    const issuesResult = yield* tracker.listIssues().pipe(
      Effect.catchAll((err) => {
        const errorMsg = String(err)
        return Effect.gen(function* () {
          yield* Effect.logError(`Poll failed: ${errorMsg}`)
          yield* PubSub.publish(pubsub, {
            _tag: "PollFailed",
            error: errorMsg,
          } satisfies SymphonyEvent)
          return [] as Issue[]
        })
      }),
    )

    const currentState = yield* Ref.get(stateRef)
    const candidates = filterCandidates(issuesResult, currentState, config)
    const slots = Math.max(config.maxConcurrent - currentState.running.size, 0)
    const toDispatch = candidates.slice(0, slots)

    yield* PubSub.publish(pubsub, {
      _tag: "PollCompleted",
      issuesFound: issuesResult.length,
      candidateCount: candidates.length,
      dispatchedCount: toDispatch.length,
    } satisfies SymphonyEvent)

    for (const issue of toDispatch) {
      yield* dispatchIssue(stateRef, pubsub, config, issue, null, workflow, hooks, agentLayer)
    }
  })
}

// --- Dispatch (Section 16.4) ---

function dispatchIssue(
  stateRef: Ref.Ref<OrchestratorState>,
  pubsub: PubSub.PubSub<SymphonyEvent>,
  config: SymphonyConfig,
  issue: Issue,
  attempt: number | null,
  workflow: Workflow | null,
  hooks: Record<string, string>,
  agentLayer: Layer.Layer<Config | EventBus | TrackerService>,
): Effect.Effect<void> {
  const id = identifier(issue)

  return Effect.gen(function* () {
    const agentEffect = runAgent(issue, workflow, hooks, attempt).pipe(
      // Normal exit: schedule continuation retry (1s)
      Effect.tap((result) =>
        Effect.gen(function* () {
          yield* PubSub.publish(pubsub, {
            _tag: "IssueCompleted",
            issueNumber: id,
            costUsd: result.costUsd,
            turns: result.numTurns,
          } satisfies SymphonyEvent)

          yield* Ref.update(stateRef, (s) => {
            const running = new Map(s.running)
            running.delete(id)
            const completed = new Set(s.completed)
            completed.add(id)
            return {
              ...s,
              running,
              completed,
              tokenTotals: {
                costUsd: s.tokenTotals.costUsd + result.costUsd,
                turns: s.tokenTotals.turns + result.numTurns,
              },
              // Continuation retry: 1s delay, attempt=1
              retryQueue: [...s.retryQueue, {
                id,
                identifier: id,
                attempt: 1,
                retryAt: Date.now() + 1000,
                error: null,
                isContinuation: true,
              }],
            }
          })
        }),
      ),
      // Abnormal exit: exponential backoff retry
      Effect.tapError((err) =>
        Effect.gen(function* () {
          const nextAttempt = (attempt ?? 0) + 1
          const backoff = computeBackoff(nextAttempt, config.maxRetryBackoffMs)

          yield* PubSub.publish(pubsub, {
            _tag: "IssueFailed",
            issueNumber: id,
            error: String(err),
            retryAttempt: nextAttempt,
            retryInMs: backoff,
          } satisfies SymphonyEvent)

          yield* Ref.update(stateRef, (s) => {
            const running = new Map(s.running)
            running.delete(id)
            return {
              ...s,
              running,
              retryQueue: [...s.retryQueue, {
                id,
                identifier: id,
                attempt: nextAttempt,
                retryAt: Date.now() + backoff,
                error: String(err),
                isContinuation: false,
              }],
            }
          })
        }),
      ),
      Effect.catchAll(() => Effect.void),
      Effect.provide(agentLayer),
    )

    const fiber = yield* Effect.fork(agentEffect)
    const now = Date.now()

    yield* Ref.update(stateRef, (s) => {
      const running = new Map(s.running)
      running.set(id, { fiber, issue, startedAt: now, lastEventAt: now, attempt })
      const claimed = new Set(s.claimed)
      claimed.add(id)
      // Remove any existing retry entry for this issue
      const retryQueue = s.retryQueue.filter((r) => r.id !== id)
      return { ...s, running, claimed, retryQueue }
    })

    yield* PubSub.publish(pubsub, {
      _tag: "IssueDispatched",
      issueNumber: id,
      title: issue.title,
    } satisfies SymphonyEvent)
  })
}

// --- Retry handling (Section 16.6) ---

function processRetries(
  stateRef: Ref.Ref<OrchestratorState>,
  config: SymphonyConfig,
  pubsub: PubSub.PubSub<SymphonyEvent>,
  tracker: Tracker,
  workflow: Workflow | null,
  hooks: Record<string, string>,
  agentLayer: Layer.Layer<Config | EventBus | TrackerService>,
): Effect.Effect<void> {
  return Effect.gen(function* () {
    const now = Date.now()
    const state = yield* Ref.get(stateRef)

    const ready = state.retryQueue.filter((r) => r.retryAt <= now)
    const pending = state.retryQueue.filter((r) => r.retryAt > now)

    if (ready.length === 0) return

    yield* Ref.update(stateRef, (s) => ({ ...s, retryQueue: pending }))

    // Fetch candidates to check eligibility
    const candidates = yield* tracker.listIssues().pipe(
      Effect.catchAll(() => Effect.succeed([] as Issue[])),
    )

    for (const entry of ready) {
      const issue = candidates.find((c) => identifier(c) === entry.id)

      if (!issue) {
        // Issue not found in active candidates — release claim
        yield* Ref.update(stateRef, (s) => {
          const claimed = new Set(s.claimed)
          claimed.delete(entry.id)
          return { ...s, claimed }
        })
        continue
      }

      // Check if issue is still active
      if (!isActiveState(issue.state, config)) {
        yield* Ref.update(stateRef, (s) => {
          const claimed = new Set(s.claimed)
          claimed.delete(entry.id)
          return { ...s, claimed }
        })
        continue
      }

      // Check concurrency slots — don't increment attempt for slot exhaustion
      const currentState = yield* Ref.get(stateRef)
      if (currentState.running.size >= config.maxConcurrent) {
        yield* Ref.update(stateRef, (s) => ({
          ...s,
          retryQueue: [...s.retryQueue, {
            ...entry,
            retryAt: now + config.pollIntervalMs,
            error: "no available orchestrator slots",
          }],
        }))
        continue
      }

      // Re-dispatch
      yield* dispatchIssue(stateRef, pubsub, config, issue, entry.attempt, workflow, hooks, agentLayer)
    }
  })
}

// --- Backoff (Section 8.4) ---

/** Spec formula: min(10000 * 2^(attempt-1), max_retry_backoff_ms) */
function computeBackoff(attempt: number, maxBackoffMs: number): number {
  return Math.min(10_000 * Math.pow(2, attempt - 1), maxBackoffMs)
}
